### ISSUE TITLE
Look into AI Code suggestions

### ISSUE THREAD

**Issue by SebieF** [sacdallago/biotrainer#93]
Suggestions from Codium AI for the version 0.9.0 PR.
              
              ## PR Code Suggestions ✨

<table><thead><tr><td>Category</td><td align=left>Suggestion&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </td><td align=center>Score</td></tr><tbody><tr><td rowspan=1><strong>Security</strong></td>
<td>

<details><summary>Improve security by using <code>yaml.safe_load()</code> instead of <code>yaml.load()</code></summary>

___

**Replace the use of <code>yaml.load()</code> with <code>yaml.safe_load()</code> to avoid potential security risks <br>associated with loading arbitrary YAML content.**
    
[biotrainer/inference/inferencer.py [86]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-d18e0475c31d49fcd5f03699579bec5b893170a66eb0f83dab045ddd1825016dR86-R86)

```diff
-output_vars = yaml.load(tmp_output_file, Loader=yaml.RoundTripLoader)
+output_vars = yaml.safe_load(tmp_output_file)
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 10</b></summary>

Why: This suggestion addresses a significant security concern by replacing `yaml.load()` with `yaml.safe_load()`, which is a safer method for loading YAML content.

</details></details></td><td align=center>10

</td></tr><tr><td rowspan=1><strong>Possible bug</strong></td>
<td>

<details><summary>Correct the superclass initialization in the <code>DeeperFNN</code> class</summary>

___

**In the <code>DeeperFNN</code> class, the superclass initialization is incorrectly calling <code>super(FNN, </code><br><code>self).__init__()</code> which should be <code>super(DeeperFNN, self).__init__()</code>. This ensures that the <br><code>DeeperFNN</code> class correctly initializes its superclass.**
    
[biotrainer/models/fnn.py [38]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-4839fd2708c2f712e8b9ea652fa512c105cabf96334f5231fad2f1f9a0c9080fR38-R38)

```diff
-super(FNN, self).__init__()
+super(DeeperFNN, self).__init__()
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 10</b></summary>

Why: This suggestion addresses a critical bug in the superclass initialization of the `DeeperFNN` class, ensuring proper inheritance and functionality.

</details></details></td><td align=center>10

</td></tr><tr><td rowspan=2><strong>Performance</strong></td>
<td>

<details><summary>Optimize the removal of special tokens by using <code>np.where()</code></summary>

___

**Use list comprehension directly in the <code>np.delete()</code> function to improve code readability <br>and efficiency.**
    
[biotrainer/embedders/huggingface_transformer_embedder.py [62-63]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-6b761024ac9732a580d1ce702992904c02c19e6cd197f1a89ef293960f37401dR62-R63)

```diff
 special_tokens_mask = self._tokenizer.get_special_tokens_mask(input_id, already_has_special_tokens=True)
-embedding = np.delete(embeddings[seq_num],
-                      [index for index, mask in enumerate(special_tokens_mask) if mask != 0], axis=0)
+embedding = np.delete(embeddings[seq_num], np.where(special_tokens_mask)[0], axis=0)
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 9</b></summary>

Why: Using `np.where()` improves both readability and performance. This change makes the code more efficient and easier to understand.

</details></details></td><td align=center>9

</td></tr><tr><td>

<details><summary>Use <code>torch.no_grad()</code> to optimize inference performance</summary>

___

**Consider using <code>torch.no_grad()</code> context manager during inference to disable gradient <br>computation, which can reduce memory consumption and increase computation speed.**
    
[biotrainer/inference/inferencer.py [232]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-d18e0475c31d49fcd5f03699579bec5b893170a66eb0f83dab045ddd1825016dR232-R232)

```diff
-inference_dict = solver.inference(dataloader, calculate_test_metrics=targets is not None)
+with torch.no_grad():
+    inference_dict = solver.inference(dataloader, calculate_test_metrics=targets is not None)
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 9</b></summary>

Why: This suggestion correctly identifies a performance optimization by using `torch.no_grad()` during inference, which can reduce memory consumption and increase computation speed.

</details></details></td><td align=center>9

</td></tr><tr><td rowspan=1><strong>Robustness</strong></td>
<td>

<details><summary>Add exception handling for model state loading to manage file-related errors</summary>

___

**Implement exception handling for the <code>torch.load</code> function to manage potential errors during <br>the loading of model states, such as file not found or corrupted files.**
    
[biotrainer/solvers/solver.py [251]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-12ca19c008c5357adf672f59c77952a54f6b7a30fe515a8d547bbf42ec6f22b8R251-R251)

```diff
-state = torch.load(checkpoint_path, map_location=torch.device(self.device))
+try:
+    state = torch.load(checkpoint_path, map_location=torch.device(self.device))
+except FileNotFoundError:
+    logger.error("Checkpoint file not found.")
+    return
+except Exception as e:
+    logger.error(f"Failed to load checkpoint: {str(e)}")
+    return
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 9</b></summary>

Why: Adding exception handling for `torch.load` is a good practice to manage potential errors such as file not found or corrupted files. This enhances the robustness of the code.

</details></details></td><td align=center>9

</td></tr><tr><td rowspan=2><strong>Best practice</strong></td>
<td>

<details><summary>Improve type checking by using <code>isinstance()</code></summary>

___

**Replace the direct type checks with <code>isinstance()</code> for better type checking, especially when <br>dealing with inheritance.**
    
[biotrainer/config/config_option.py [67-68]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-6b1e20e34aafc2390d317fb619501b4b6bb825febd137af7b2d01d314502d19eR67-R68)

```diff
-return ("range" in str(self.value) or type(self.value) is list or
-        (type(self.value) is str and "[" in self.value and "]" in self.value))
+return ("range" in str(self.value) or isinstance(self.value, list) or
+        (isinstance(self.value, str) and "[" in self.value and "]" in self.value))
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 8</b></summary>

Why: Using `isinstance()` is a best practice for type checking, especially when dealing with inheritance. This change improves code robustness and readability.

</details></details></td><td align=center>8

</td></tr><tr><td>

<details><summary>Use more specific exception types for clearer error handling</summary>

___

**Use a more specific exception type than the general <code>Exception</code> to provide clearer error <br>handling.**
    
[biotrainer/config/configurator.py [81-82]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-6612eb900eb1575da47046296ca2d5a17103e842d5c9d6d17660c0be0d92612fR81-R82)

```diff
-except Exception as e:
-    raise Exception(f"Loading {embedder_name} automatically and as {tokenizer_class.__class__.__name__} failed!"
-                    f" Please provide a custom_embedder script for your use-case.") from e
+except ImportError as e:
+    raise ImportError(f"Loading {embedder_name} automatically and as {tokenizer_class.__class__.__name__} failed!"
+                      f" Please provide a custom_embedder script for your use-case.") from e
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 7</b></summary>

Why: Using a more specific exception type like `ImportError` provides clearer error handling and makes the code easier to debug. However, the improvement is minor and context-specific.

</details></details></td><td align=center>7

</td></tr><tr><td rowspan=4><strong>Enhancement</strong></td>
<td>

<details><summary>Enhance error messages for clarity and debugging</summary>

___

**Replace the manual exception raising for unknown <code>split_name</code> with a more informative error <br>message that includes the available splits.**
    
[biotrainer/inference/inferencer.py [191]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-d18e0475c31d49fcd5f03699579bec5b893170a66eb0f83dab045ddd1825016dR191-R191)

```diff
-raise Exception(f"Unknown split_name {split_name} for given configuration!")
+if split_name not in self.solvers_and_loaders_by_split:
+    available_splits = ', '.join(self.solvers_and_loaders_by_split.keys())
+    raise ValueError(f"Unknown split_name '{split_name}'. Available splits are: {available_splits}")
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 8</b></summary>

Why: The suggestion improves the clarity of error messages by including available split names, which aids in debugging and provides more informative feedback to the user.

</details></details></td><td align=center>8

</td></tr><tr><td>

<details><summary>Apply dropout consistently to both feature and attention convolutions in the <code>LightAttention</code> class</summary>

___

**In the <code>LightAttention</code> class, the <code>dropout</code> operation is applied only to the output of <br><code>feature_convolution</code> but not to <code>attention_convolution</code>. Consistently applying dropout to <br>both could potentially improve model performance by regularizing both features and <br>attention mechanisms.**
    
[biotrainer/models/light_attention.py [47]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-6df65b193c9e38d99a35caca258ad9cba26adbe51bf1b521b696f87aecf26551R47-R47)

```diff
 o = self.dropout(o)
+attention = self.dropout(attention)
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 8</b></summary>

Why: This suggestion potentially improves model performance by regularizing both features and attention mechanisms, making it a valuable enhancement.

</details></details></td><td align=center>8

</td></tr><tr><td>

<details><summary>Enhance the <code>_early_stop</code> method by logging the reason for stopping</summary>

___

**Modify the <code>_early_stop</code> method to log the reason for stopping, which could be due to <br>achieving a new minimum loss or reaching the patience limit. This enhances debugging and <br>monitoring capabilities.**
    
[biotrainer/solvers/solver.py [299]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-12ca19c008c5357adf672f59c77952a54f6b7a30fe515a8d547bbf42ec6f22b8R299-R299)

```diff
 if self._stop_count == 0:
+    logger.info("Early stopping due to patience limit reached.")
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 8</b></summary>

Why: Logging the reason for early stopping enhances debugging and monitoring capabilities, making it easier to understand why the training was stopped. This is a useful enhancement for tracking the training process.

</details></details></td><td align=center>8

</td></tr><tr><td>

<details><summary>Improve variable naming for clarity in the <code>FNN</code> class's <code>forward</code> method</summary>

___

**In the <code>FNN</code> class, consider using a more descriptive variable name for the input tensor <code>x</code> <br>in the <code>forward</code> method. Renaming <code>x</code> to <code>input_tensor</code> would improve code readability and make <br>the method's purpose clearer.**
    
[biotrainer/models/fnn.py [20]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-4839fd2708c2f712e8b9ea652fa512c105cabf96334f5231fad2f1f9a0c9080fR20-R20)

```diff
-def forward(self, x):
+def forward(self, input_tensor):
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 5</b></summary>

Why: While this suggestion enhances code readability, it is a minor improvement and does not address any functional issues.

</details></details></td><td align=center>5

</td></tr><tr><td rowspan=4><strong>Maintainability</strong></td>
<td>

<details><summary>Simplify dictionary creation from iterable using list comprehension</summary>

___

**Use list comprehension to simplify the creation of <code>embeddings_dict</code> from <code>embeddings</code> when it <br>is not a dictionary.**
    
[biotrainer/inference/inferencer.py [278]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-d18e0475c31d49fcd5f03699579bec5b893170a66eb0f83dab045ddd1825016dR278-R278)

```diff
-embeddings_dict = {str(idx): embedding for idx, embedding in enumerate(embeddings)}
+embeddings_dict = dict(enumerate(embeddings)) if not isinstance(embeddings, Dict) else embeddings
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 7</b></summary>

Why: This suggestion simplifies the code for creating `embeddings_dict` from an iterable, improving code readability and maintainability. However, the improvement is minor.

</details></details></td><td align=center>7

</td></tr><tr><td>

<details><summary>Refactor the mask calculation into a separate method in the <code>LightAttention</code> class</summary>

___

**The <code>mask</code> calculation in the <code>forward</code> method of the <code>LightAttention</code> class should be moved to <br>a separate method to improve code readability and maintainability. This change will make <br>the <code>forward</code> method cleaner and focus primarily on the forward pass logic.**
    
[biotrainer/models/light_attention.py [43]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-6df65b193c9e38d99a35caca258ad9cba26adbe51bf1b521b696f87aecf26551R43-R43)

```diff
-mask = x.sum(dim=-1) != utils.SEQUENCE_PAD_VALUE
+mask = self.calculate_mask(x)
 
+def calculate_mask(self, x):
+    return x.sum(dim=-1) != utils.SEQUENCE_PAD_VALUE
+
```
                   

<details><summary><b>Suggestion importance[1-10]: 7</b></summary>

Why: This suggestion improves code readability and maintainability by separating concerns, but it does not address a critical issue.

</details></details></td><td align=center>7

</td></tr><tr><td>

<details><summary>Refactor to separate training and validation into distinct methods for better modularity</summary>

___

**Refactor the <code>train</code> method to separate the training and validation phases into their own <br>methods. This improves code readability and maintainability by modularizing the training <br>process.**
    
[biotrainer/solvers/solver.py [66]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-12ca19c008c5357adf672f59c77952a54f6b7a30fe515a8d547bbf42ec6f22b8R66-R66)

```diff
 for epoch in range(self.start_epoch, self.number_of_epochs):
+    self._train_epoch(training_dataloader, epoch)
+    self._validate_epoch(validation_dataloader, epoch)
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 7</b></summary>

Why: Refactoring the `train` method to separate training and validation phases improves code readability and maintainability. However, this suggestion requires additional implementation details for the new methods, which are not provided.

</details></details></td><td align=center>7

</td></tr><tr><td>

<details><summary>Simplify dictionary initialization using comprehension</summary>

___

**Use dictionary comprehension to simplify the initialization of <code>__DATASETS</code> and <br><code>__COLLATE_FUNCTIONS</code>.**
    
[biotrainer/datasets/__init__.py [9-22]](https://github.com/sacdallago/biotrainer/pull/92/files#diff-931f0ce7103dd6457b0ab4e66fe2dd5aa194c46ae70cedbc5977d37e109676d5R9-R22)

```diff
-__DATASETS = {
-    Protocol.residue_to_class: ResidueEmbeddingsClassificationDataset,
-    Protocol.residues_to_class: ResidueEmbeddingsClassificationDataset,
-    Protocol.residues_to_value: ResidueEmbeddingsRegressionDataset,
-    Protocol.sequence_to_class: SequenceEmbeddingsClassificationDataset,
-    Protocol.sequence_to_value: SequenceEmbeddingsRegressionDataset,
-}
-__COLLATE_FUNCTIONS = {
-    Protocol.residue_to_class: pad_residue_embeddings,
-    Protocol.residues_to_class: pad_residues_embeddings,
-    Protocol.residues_to_value: pad_residues_embeddings,
-    Protocol.sequence_to_class: pad_sequence_embeddings,
-    Protocol.sequence_to_value: pad_sequence_embeddings,
-}
+__DATASETS = {protocol: dataset for protocol, dataset in zip(
+    [Protocol.residue_to_class, Protocol.residues_to_class, Protocol.residues_to_value,
+     Protocol.sequence_to_class, Protocol.sequence_to_value],
+    [ResidueEmbeddingsClassificationDataset, ResidueEmbeddingsClassificationDataset, ResidueEmbeddingsRegressionDataset,
+     SequenceEmbeddingsClassificationDataset, SequenceEmbeddingsRegressionDataset])}
+__COLLATE_FUNCTIONS = {protocol: function for protocol, function in zip(
+    [Protocol.residue_to_class, Protocol.residues_to_class, Protocol.residues_to_value,
+     Protocol.sequence_to_class, Protocol.sequence_to_value],
+    [pad_residue_embeddings, pad_residues_embeddings, pad_residues_embeddings,
+     pad_sequence_embeddings, pad_sequence_embeddings])}
 
```
                   

<details><summary><b>Suggestion importance[1-10]: 6</b></summary>

Why: While dictionary comprehension can make the code more concise, it may also reduce readability for some developers. The improvement is more about code style and maintainability.

</details></details></td><td align=center>6

</td></tr></tr></tbody></table>

_Originally posted by @CodiumAI-Agent in https://github.com/sacdallago/biotrainer/issues/92#issuecomment-2171411737_
